org.osgi.test.junit5/src/main/java/org/osgi/test/junit5/service/ServiceUseExtension.java

```diff
@@ -108,33 +120,36 @@ public Builder(Class<T> serviceType, BundleContextExtension contextExtension) {
 	final static Namespace					NAMESPACE	= Namespace.create(ServiceUseExtension.class);
 
 	private final BundleContextExtension	contextExtension;
-	private final TrackServices<T>			trackServices;
+	private final Filter					filter;
+	private final int						cardinality;
+	private final long						timeout;
+	private volatile TrackServices<T>		trackServices;
 
 	@SuppressWarnings("unchecked")
 	protected ServiceUseExtension(Class<T> serviceType, BundleContextExtension contextExtension, Filter filter,
 		int cardinality, long timeout) {
 		super(serviceType);
 		this.contextExtension = contextExtension;
-		this.trackServices = new TrackServices<>(filter, cardinality, timeout);
+		this.filter = filter;
+		this.cardinality = cardinality;
+		this.timeout = timeout;
 	}
 
 	@Override
 	public void beforeEach(ExtensionContext extensionContext) throws Exception {
 		contextExtension.beforeEach(extensionContext);
-		trackServices.init(contextExtension.getBundleContext(extensionContext));
-		injectFields(extensionContext, extensionContext.getRequiredTestInstance(), ReflectionUtils::isNotStatic);
-	}
-
-	@Override
-	public void afterEach(ExtensionContext extensionContext) throws Exception {
-		close();
+		// set this in order to be able to honour the BaseServiceUse contract
+		trackServices = getTrackServices(extensionContext);
```

Comment:
Do we even need to honour the full `BaseServiceUse` contract? If we inject an instance of `ServiceAware<T>` instead (either as a parameter or as an instance field in the test), then the extension itself doesn't need to implement `ServiceAware<T>`. That's how they seem to have done it in `SoftAssertionsExtension`, and in fact all the extensions that I've seen - the old pattern with rules (where you access members of the rule directly) has gone by the wayside in favour of injecting state objects into the tests.

If that's too hard, maybe we can use it as-is and see how we go.

Reply:
I'm going to approve as-is because I think this is at the least an improvement, even if not the final solution.